pandas/_libs/lib.pyx

```diff
@@ -3038,3 +3038,25 @@ def is_bool_list(obj: list) -> bool:
 
     # Note: we return True for empty list
     return True
+
+
+def dtypes_all_equal(list types not None) -> bool:
```

Comment:
i am really puzzled by this type signature

Reply:
That's cython syntax to indicate the argument will never be None (see eg https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html?highlight=%22not%20None%22#syntax, although that section is about memoryviews, that applies in general)

Reply:
NBD I don’t think we use the “not None” pattern elsewhere.

BTW my internet is down so I’ll be unresponsive for a bit.

Reply:
Yeah, we use it in pyarrow, but indeed don't seem to use it in pandas at the moment. It's standard cython, and it reduces the overhead a bit (but maybe not enough to deviate, from a quick check it's around 100 ns, and to be clear it still errors when passing None, it will only do that once upfront, instead of afterwards on every interaction with the list)

Reply:
got it

Reply:
> from a quick check it's around 100 ns

does this imply that adding `not None` to other cython functions could save 100ns elsewhere?

Reply:
I suppose it depends on the actual type and what operation you do with it (eg here we did a getitem into that list, and the generated C would check for None to raise a "None is not subscriptable", as you would get in normal python), but generally yes. I also differ function by function whether this is worth it, and I don't know if you can achieve the same with the python-style annotations.